paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/SparkWrite.java

```diff
@@ -75,6 +114,34 @@ public InsertableRelation toInsertableRelation() {
         };
     }
 
+    private boolean isDynamicBucketTable() {
+        return table.isFileStoreTable()
+                && ((FileStoreTable) table).bucketMode() == BucketMode.DYNAMIC;
+    }
+
+    private static Iterator<InternalRow> assignBucket(
+            FileStoreTable fileStoreTable, BatchWriteBuilder writeBuilder, Iterator<Row> rows) {
+        long targetBucketRowNumber = fileStoreTable.coreOptions().dynamicBucketTargetRowNum();
+        List<InternalRow> dynamicBucketRows = new ArrayList<>();
```

Comment:
I think it is better to return a lazy Iterator for Spark `RDD.mapPartitions`, otherwise there is a risk of OOM